docs/Execution.md

```diff
@@ -0,0 +1,107 @@
+    };
+} QUIC_LISTENER_EVENT;
+
+typedef
+_IRQL_requires_max_(PASSIVE_LEVEL)
+_Function_class_(QUIC_LISTENER_CALLBACK)
+QUIC_STATUS
+(QUIC_API QUIC_LISTENER_CALLBACK)(
+    _In_ HQUIC Listener,
+    _In_opt_ void* Context,
+    _Inout_ QUIC_LISTENER_EVENT* Event
+    );
+```
+
+Above is an example of a callback delivered to the listener interface.
+The application must register a callback handler to manage all events MsQuic may indicate, returning a status to show if it was successfully handled or not.
+
+This approach differs significantly from BSD sockets, where the application must make a call (e.g., `send` or `recv`) to determine if something happened.
+This design choice was made for several reasons:
+
+- The MsQuic API **runs in-process**, eliminating the need for a kernel to user mode boundary switch to notify the application layer. This makes the callback-based design more practical compared to BSD sockets.
+
+- MsQuic, due to the QUIC protocol, has numerous event types. Applications may have hundreds of objects with potential state changes. The callback model allows the application to avoid managing pending calls on each object.
+
+- Writing correct, performant code on top of the BSD-style interface has proven challenging. Callbacks, executed at the correct time and on the correct thread/processor, enable MsQuic to abstract much complexity from applications, making things "just work" out of the box.
+
+- It simplifies MsQuic's logic by eliminating the need for a queue or cached state to indicate to the application. In the BSD model, the networking stack must wait for a top-down call from the application before indicating completion, increasing code size, complexity, and memory usage.
+
+### Writing Event Handlers
+
+Event handlers are **essential** for all objects that support them, as much of the MsQuic API operates through these callbacks.
+Critical events, such as "shutdown complete" notifications, provide vital information necessary for the application to function correctly.
+Without these events, the application cannot determine when it is safe to clean up objects.
+
+Applications should keep the execution time within callbacks **to a minimum**.
+MsQuic does not use separate threads for protocol execution and upcalls to the application.
+Therefore, any significant delays in the callback **will delay the protocol**.
+Any substantial work required by the application should be performed on its own thread.
+
+This does not mean the application cannot perform any work in the callback handler.
+In fact, many operations are designed to be most efficient when executed within the callback.
+For example, closing a handle to a connection or stream is ideally done during the "shutdown complete" indication.
+
+A crucial aspect of this design is that all blocking API (down) calls invoked within a callback always occur inline (to prevent deadlocks) and will take precedence over any calls in progress or queued from a separate thread.
```

Comment:
so an application must be very careful if they are doing some work inline in their callback, and some work deferred in their own separate threads? that sounds difficult to accomplish without locks involved to guarantee predictable order-of-execution. and it sounds like acquiring locks in the callback is going to be a bad idea - as that could cause effectively random latency jumps on a connection :(

So is the model more often, "Pick one: handle inline, or defer work to your own threads; if you do both, you'll need to provide your own synchronization since any calls inline will be preferred over whatever your own threads are doing"?

Reply:
even synchronizing won't help the app track exactly what is happening in what order, since even if I synchronize "I'm calling QUIC function X from my own thread; now my lock unblocks the callback thread and I compare state and make a new call from my callback -- oops, my call from my callback just took effect before my prior call from my own thread!"

Reply:
Locks by themselves are not a bad thing. I believe most production code on top of MsQuic uses locks just fine in the callbacks today. Locks that you hold or contend a lot are bad. But that's true in the general sense; not just in our callbacks.

The only really bad 'gotcha' we have is if you try to make a blocking call on Connection B in the callback of Connection A. This can results in a deadlock and is not super obvious to the developer at first glance.

So, generally, my recommendation is to run code in the callback if it's purely computational (i.e. no I/O). But anything that might block on some other component or take a really long time, you should queue a worker.